File: Aero Linea/Aerolinea.API/Aerolinea.API/Helpers/AgenciaAuthMiddleware.cs
using Aerolinea.API.Repositories;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Filters;

namespace Aerolinea.API.Helpers
{
    /// <summary>
    /// Representa la identidad de una agencia externa autenticada.
    /// Contiene los datos basicos que se propagan en el contexto HTTP tras validar el token.
    /// </summary>
    public class AgenciaIdentidad
    {
        public int ID { get; set; }
        public string Nombre { get; set; } = string.Empty;
        public string URLAgencia { get; set; } = string.Empty;
    }

    /// <summary>
    /// Filtro de autenticacion para agencias externas que consumen la API.
    /// Valida el token enviado en el header X-Agencia-Token antes de ejecutar el action del controller.
    /// Si el token es invalido o no existe, detiene la ejecucion y retorna 401 Unauthorized.
    /// </summary>
    public class AgenciaAuthMiddleware : IAsyncActionFilter
    {
        private readonly AgenciaRepository _repo;

        /// <summary>
        /// Inicializa el filtro con el repositorio de agencias necesario para validar el token.
        /// </summary>
        public AgenciaAuthMiddleware(AgenciaRepository repo)
        {
            _repo = repo;
        }

        /// <summary>
        /// Ejecuta la validacion del token de agencia antes de que el action del controller sea invocado.
        /// Extrae el token del header X-Agencia-Token, consulta la base de datos para verificarlo
        /// y almacena los datos de la agencia en HttpContext.Items si es valido.
        /// Retorna 401 si el token esta ausente o no corresponde a ninguna agencia registrada.
        /// </summary>
        public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
        {
            var token = context.HttpContext.Request.Headers["X-Agencia-Token"].FirstOrDefault();
            if (string.IsNullOrEmpty(token))
            {
                context.Result = new UnauthorizedObjectResult(new { message = "token de agencia requerido" });
                return;
            }

            var agencia = await _repo.ObtenerAgenciaPorToken(token);
            if (agencia == null)
            {
                context.Result = new UnauthorizedObjectResult(new { message = "token inválido — agencia no reconocida" });
                return;
            }

            context.HttpContext.Items["agencia_id"] = agencia.ID;
            context.HttpContext.Items["agencia_nombre"] = agencia.Nombre;
            context.HttpContext.Items["agencia_url"] = agencia.URLAgencia;

            await next();
        }
    }
}

File: Cadena de Hoteles/Hotel/src/main/java/org/example/helpers/AgenciaAuthMiddleware.java
package org.example.helpers;

import io.javalin.http.Context;
import io.javalin.http.Handler;
import org.example.dtos.AgenciaIdentidad;
import org.example.repositories.AgenciaRepository;
import java.util.Map;

/**
 * Middleware de autenticacion para rutas protegidas de agencias externas.
 * Valida el token enviado en el header X-Agencia-Token e inyecta
 * la identidad de la agencia en el contexto de la peticion.
 */
public class AgenciaAuthMiddleware {

    private static final AgenciaRepository repo = new AgenciaRepository();

    /**
     * Verifica el token de agencia en el header de la peticion.
     * Si el token es valido, inyecta el ID, nombre y URL de la agencia
     * como atributos del contexto para que los controllers puedan usarlos.
     * Debe llamarse al inicio de cada ruta protegida de agencia.
     *
     * @param ctx contexto de la peticion HTTP de Javalin.
     * @return true si el token es valido y la agencia fue identificada;
     *         false si el token falta o no corresponde a ninguna agencia,
     *         en cuyo caso ya se escribe la respuesta 401 en el contexto.
     */
    public static boolean verificar(Context ctx) {
        String token = ctx.header("X-Agencia-Token");

        if (token == null || token.isBlank()) {
            ctx.status(401).json(Map.of("mensaje", "Token de agencia requerido"));
            return false;
        }

        AgenciaIdentidad agencia = repo.obtenerAgenciaPorToken(token);
        if (agencia == null) {
            ctx.status(401).json(Map.of("mensaje", "Token inválido — agencia no reconocida"));
            return false;
        }

        // Inyectar identidad en el contexto, igual que el AuthMiddleware de sesiones
        ctx.attribute("agenciaId",     agencia.getId());
        ctx.attribute("agenciaNombre", agencia.getNombre());
        ctx.attribute("agenciaUrl",    agencia.getUrlAgencia());
        return true;
    }
}